server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -193,33 +193,38 @@ public StoredContext newEmptySystemContext() {
      */
     public StoredContext newTraceContext() {
         final ThreadContextStruct originalContext = threadLocal.get();
-        final Map<String, String> newRequestHeaders = new HashMap<>(originalContext.requestHeaders);
-        final Map<String, Object> newTransientHeaders = new HashMap<>(originalContext.transientHeaders);
+        // this is the context when this method returns
+        final ThreadContextStruct newContext;
+        if (originalContext.hasTraceContext() == false) {
```

Comment:
This assumes that these headers are all either totally absent or totally present, which I think is the case, but deserves an assertion somewhere.

Also could this be a separate PR (maybe to go in first?)

Reply:
Sure, I'll do 👍

Reply:
@DaveCTurner following up on this next. 
>This assumes that these headers are all either totally absent or totally present

`hasTraceContext` requires just one of the trace headers to be present, and if that's the case we clean request headers. If tracing is disabled (using the noop tracer), TRACE_STATE and APM_TRACE_CONTEXT.

TRACE_PARENT_HTTP_HEADER is also the only of the headers that's required according to the [w3 spec](https://www.w3.org/TR/trace-context-1/#design-overview). Want to assert this one exists if any of the other ones is present?

Reply:
Ah wait sorry I misread what `hasTraceContext` does, it's a disjunction. This is fine as-is.